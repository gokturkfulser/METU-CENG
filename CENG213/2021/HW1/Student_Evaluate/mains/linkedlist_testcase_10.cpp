#include <iostream>

#include "LinkedList.h"
//#include "../LinkedList.h"

template<class T>
void printSizeIncludingDummies(LinkedList<T> *llist) {
    std::cout << "-> Calculating the size including dummies." << std::endl;

    int count = 0;

    Node<T> *temp = llist->getDummyHead();

    while (temp) {
        count += 1;
        temp = temp->next;
    }

    std::cout << "size is " << count << std::endl;
}

template<class T>
void traverseReverse(LinkedList<T> *llist) {
    std::cout << "-> Printing the list in reverse." << std::endl;

    if (llist == NULL) {
        std::cout << "NULL" << std::endl;
        return;
    } else if (llist->isEmpty()) {
        std::cout << "The list is empty." << std::endl;
        return;
    } else {
        Node<T> *node = llist->getActualTail();

        while (node && node->prev) {
            std::cout << *node << std::endl;
            node = node->prev;
        }
    }
}

/*
 * Case 10 : Default constructor; various inserts; delete middle node; getSize and isEmpty; print.
 */
int main() {
    std::cout << "-> Creating a linked list with default constructor." << std::endl;

    LinkedList<int> llist;

    std::cout << "-> Making some insertions to the list." << std::endl;

    llist.insertAtTheBeginning(32);
    llist.insertAtTheEnd(34);
    llist.insertAtTheBeginning(31);
    llist.insertAtTheEnd(35);
    llist.insertBeforeGivenNode(33, llist.getNode(34));

    std::cout << "-> Printing the list." << std::endl;

    llist.traverse();
    traverseReverse(&llist);
    printSizeIncludingDummies(&llist);

    std::cout << "-> Deleting the middle node from the list and then printing the list." << std::endl;

    llist.deleteNode(llist.getNode(33));

    llist.traverse();
    traverseReverse(&llist);
    printSizeIncludingDummies(&llist);

    std::cout << "-> Calculating the size." << std::endl;

    std::cout << "size is " << llist.getSize() << std::endl;

    std::cout << "-> Checking if the list is empty or not." << std::endl;

    if (llist.isEmpty()) {
        std::cout << "linked list is empty" << std::endl;
    } else {
        std::cout << "linked list is not empty" << std::endl;
    }

    return 0;
}
