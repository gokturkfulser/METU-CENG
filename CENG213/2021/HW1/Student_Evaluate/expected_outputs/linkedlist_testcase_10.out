-> Creating a linked list with default constructor.
-> Making some insertions to the list.
-> Printing the list.
31
32
33
34
35
-> Printing the list in reverse.
35
34
33
32
31
-> Calculating the size including dummies.
size is 7
-> Deleting the middle node from the list and then printing the list.
31
32
34
35
-> Printing the list in reverse.
35
34
32
31
-> Calculating the size including dummies.
size is 6
-> Calculating the size.
size is 4
-> Checking if the list is empty or not.
linked list is not empty
